### PR TITLE
VZ-7156: Verrazzano should do a rolling restart of WebLogic when Istio version changes

### PR DESCRIPTION
**Pull request by Abhijithang**
This PR provides fixes for the following issues: 
- Do a rolling restart of the WebLogic domain when the Istio version changes with a skew of 2 minor versions maximum. 
- Stops the WebLogic domain completely before the upgrade if the Istio version changes where the skew is more than two minor versions and starts after the upgrade. 
- Fix the issue where the WebLogic domain should not restart if the Istio version didn't change.

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go [hunk 1]
```diff
@@ -34,6 +34,11 @@ func RestartApps(log vzlog.VerrazzanoLogger, client clipkg.Client, generation in
 		return err
 	}
 
+	// Rolling restart Weblogic domain pods if the Istio version skew is 2 minor versions max
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/app_restart.go:37]: Comment show say "Stop domain if max skew > 2, rolling restart cannot be done in this case".  Or something like that.

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go [hunk 2]
```diff
@@ -70,6 +75,37 @@ func StopDomainsUsingOldEnvoy(log vzlog.VerrazzanoLogger, client clipkg.Client)
 	return nil
 }
 
+// RestartDomainsUsingOldEnvoyMaxSkewTwoMinorVersions restarts all the WebLogic domains using Envoy 1.7.3
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/app_restart.go:78]: comment is wrong
using Envoy 1.7.3

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go [hunk 3]
```diff
@@ -70,6 +75,37 @@ func StopDomainsUsingOldEnvoy(log vzlog.VerrazzanoLogger, client clipkg.Client)
...
+func RestartDomainsUsingOldEnvoyMaxSkewTwoMinorVersions(log vzlog.VerrazzanoLogger, client clipkg.Client, restartVersion string) error {
+	// Generate a restart version that will not change for this Verrazzano version
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/app_restart.go:80]: remove these 2 comments

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go [hunk 4]
```diff
@@ -70,6 +75,37 @@ func StopDomainsUsingOldEnvoy(log vzlog.VerrazzanoLogger, client clipkg.Client)
...
+	//restartVersion := "upgrade-" + strconv.Itoa(int(generation))
+
+	// Get the latest Istio proxy image name from the bom
+	istioProxyImage, err := getIstioProxyImageFromBom()
+	if err != nil {
+		return log.ErrorfNewErr("Failed, restart components cannot find Istio proxy image in BOM: %v", err)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/app_restart.go:86]: remove "restart components "

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go (hunk 2 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/component/istio/app_restart.go:78]: Mention "rolling restart"

Comment (Abhijithang) [reply platform-operator/controllers/verrazzano/component/istio/app_restart.go:78]: Sorry, I missed it. Will do the change.

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go (hunk 3 above)
Comment (Abhijithang) [reply platform-operator/controllers/verrazzano/component/istio/app_restart.go]: Done.

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go (hunk 4 above)
Comment (Abhijithang) [reply platform-operator/controllers/verrazzano/component/istio/app_restart.go]: Sorry, removed them.

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go [hunk 5]
```diff
@@ -94,14 +127,46 @@ func stopDomainIfNeeded(log vzlog.VerrazzanoLogger, client clipkg.Client, appCon
...
+	selector = selector.Add(*weblogicReq).Add(*compReq).Add(*appConfNameReq)
+
+	// Get the pods using the label selector
+	podList, err := goClient.CoreV1().Pods(appConfig.Namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: selector.String()})
+	if err != nil {
+		return log.ErrorfNewErr("Failed to list pods for Domain %s/%s: %v", appConfig.Namespace, wlName, err)
+	}
+
+	// Check if weblogic domain pods contain the old Istio proxy image where skew is 2 minor versions at max.
+	found := !DoesPodContainOldIstioSidecarSkewGreaterThanTwoMinorVersion(log, podList, "OAM WebLogic Domain", wlName, istioProxyImage)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/app_restart.go:162]: This Does.. function has the following comment which is wrong
log.Oncef("Restarting %s %s which has a pod with an old Istio proxy with a major version change%s", workloadType, workloadName,

Comment (Abhijithang) [reply platform-operator/controllers/verrazzano/component/istio/app_restart.go:162]: Sorry again. Refactored it correctly now. Thanks.

File: platform-operator/controllers/verrazzano/component/istio/app_restart.go (hunk 2 above)
Comment (pshirodkar-orcl) [reply platform-operator/controllers/verrazzano/component/istio/app_restart.go]: Aren't we using envoy 1.14.3?

**Review by pfmackin — APPROVED** [review]

**Pull request merged**